gio-macros/src/lib.rs

```diff
@@ -0,0 +1,63 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+extern crate proc_macro;
+
+mod action_map;
+
+use proc_macro::TokenStream;
+use syn::{parse_macro_input, ItemImpl};
+
+/// Macro for creating [`gio::Action`]s and registering them in a given [`gio::ActionMap`].
+/// It generates a method `register_actions` with a following signature:
+///
+/// ```rust
+/// # struct X; impl X {
+/// fn register_actions<AM: glib::object::IsA<gio::ActionMap>>(&self, map: &AM)
+/// # {} }
+/// ```
+///
+/// # Example
+/// ```rust
+/// use gio::prelude::*;
+///
+/// #[derive(glib::Downgrade)]
+/// pub struct MyApplication(gio::Application);
+///
+/// impl MyApplication {
+///     pub fn new() -> Self {
+///         let app = Self(gio::Application::new(
+///             Some("com.example.MyApplication"),
+///             gio::ApplicationFlags::FLAGS_NONE,
+///         ));
+///         app.register_actions(&app.0);
+///         app
+///     }
+/// }
+///
+/// #[gio::actions]
```

Comment:
What I was actually thinking of is to rather uses this with subclasses, and have the gio::actions attribute implement ActionMap/ActionGroup for that subclass automatically and then you can add the actions directly to self as it would be an action map.

Reply:
You still may use this macro with subclasses. Just invoke `self.register_actions(&self)` in a constructor method.

Reply:
Sure, I guess we can provide another derive macro like the CompositeTemplate for such use case I guess.

Reply:
@bilelmoussaoui Frankly, I was under impression that Gtk is moving [away from inheritance](https://blog.gtk.org/2020/04/23/custom-widgets-in-gtk-4-introduction/).

> The general direction of our API changes has been to emphasize delegation over subclassing.

>  many widgets are now final classes

Reply:
> @bilelmoussaoui Frankly, I was under impression that Gtk is moving [away from inheritance](https://blog.gtk.org/2020/04/23/custom-widgets-in-gtk-4-introduction/).
> 
> > The general direction of our API changes has been to emphasize delegation over subclassing.
> 
> > many widgets are now final classes

You're mis-understanding what the announcement meant I guess. in gtk3 creating a custom widget was a painful process which was simplified in gtk4, and gtk developers expect you to create custom widgets (a subclass of gtk::Widget) where you compose multiple pre-defined widgets.

Reply:
Without subclassing you can't use composite templates, properties, signals and all what gtk offers :)

Reply:
> i Frankly, I was under impression that Gtk is moving [away from inheritance](https://blog.gtk.org/2020/04/23/custom-widgets-in-gtk-4-introduction/).

It's more of "everything now will be inheriting from `GtkWidget` directly rather than 4 different classes deep"